core/deadline.go

```diff
@@ -207,18 +206,37 @@ func NewDutyDeadlineFunc(ctx context.Context, eth2Cl eth2wrap.Client) (DeadlineF
 	}
 
 	return func(duty Duty) (time.Time, bool) {
-		if duty.Type == DutyExit || duty.Type == DutyBuilderRegistration {
+		switch duty.Type {
+		case DutyExit, DutyBuilderRegistration:
 			// Do not timeout exit or registration duties.
 			return time.Time{}, false
+		default:
 		}
 
-		start := genesisTime.Add(slotDuration * time.Duration(duty.Slot))
-		delta := slotDuration * time.Duration(lateFactor)
-		if delta < lateMin {
-			delta = lateMin
+		var (
+			start    = genesisTime.Add(slotDuration * time.Duration(duty.Slot))
+			margin   = slotDuration / marginFactor
+			duration time.Duration
+		)
+
+		switch duty.Type {
+		case DutyProposer:
+			duration = slotDuration / 3
+		case DutyAttester, DutyAggregator, DutyPrepareAggregator:
+			// Even though attestations and aggregations are acceptable even after 2 slots, the rewards are heavily diminished.
+			duration = 2 * slotDuration
+		case DutySyncMessage:
+			duration = 2 * slotDuration / 3
+		case DutySyncContribution, DutyPrepareSyncContribution:
+			duration = slotDuration
+		case DutyRandao:
+			// Randao should be accepted as long as it is not after the proposer's slot.
+			// However, given how cheap the operation is, the beacon node failing to provide it for 2 slots signals that there is something wrong with the beacon node itself.
+			duration = 2 * slotDuration
+		default:
+			duration = slotDuration
 		}
-		end := start.Add(delta)
 
-		return end, true
+		return start.Add(duration + margin), true
```

Comment:
Not sure we should add margin this way, as it adds it to DutyProposer, DutyRandao, and DutySyncMessage. These should all expire within ~4 seconds (maybe 6?), not 28 seconds.

Reply:
Margin is 0.5s for Ethereum: slot_duration/24.

Reply:
Regarding Randao, me and @KaloyanTanev discussed this 1:1 and agreed to give it a longer window of two slots.

Reply:
Fixed timings for DutyProposer, DutyRandao, and DutySyncMessage.